daemon/cmd/fqdn.go

```diff
@@ -400,3 +321,124 @@ func (d *Daemon) notifyOnDNSMsg(
+	//
+	// We do not do a `defer unlock()` here, as we should release the lock before
+	// doing final bookkeeping.
+	mutexAcquireStart := time.Now()
+	d.dnsNameManager.LockName(qname)
+
+	if d := time.Since(mutexAcquireStart); d >= option.Config.DNSProxyLockTimeout {
+		log.WithFields(logrus.Fields{
+			logfields.DNSName:  qname,
+			logfields.Duration: d,
+			logfields.Expected: option.Config.DNSProxyLockTimeout,
+		}).Warnf("Name lock acquisition time took longer than expected. "+
+			"Potentially too many parallel DNS requests being processed, "+
+			"consider adjusting --%s and/or --%s",
+			option.DNSProxyLockCount, option.DNSProxyLockTimeout)
+	}
+
+	logDebug := log.Logger.IsLevelEnabled(logrus.DebugLevel)
+	if logDebug {
+		log.WithField(logfields.EndpointID, ep.ID).Debug("Recording DNS lookup in endpoint specific cache")
+	}
+	// This must happen before the NameManager update below, to ensure that
+	// this data is included in the serialized Endpoint object.
+	// We also need to add to the cache before we purge any matching zombies
+	// because they are locked separately and we want to keep the allowed IPs
+	// consistent if a regeneration happens between the two steps. If an update
+	// doesn't happen in the case, we play it safe and don't purge the zombie
+	// in case of races.
+	if updated := ep.DNSHistory.Update(lookupTime, qname, responseIPs, int(TTL)); updated {
+		ep.DNSZombies.ForceExpireByNameIP(lookupTime, qname, responseIPs...)
+		ep.SyncEndpointHeaderFile()
+	}
+
+	if logDebug {
+		log.WithFields(logrus.Fields{
+			"qname": qname,
+			"ips":   responseIPs,
+		}).Debug("Updating DNS name in cache from response to query")
+	}
+
+	updateCtx, updateCancel := context.WithTimeout(d.ctx, option.Config.FQDNProxyResponseMaxDelay)
+	defer updateCancel()
+	updateStart := time.Now()
+
+	dpUpdates := d.dnsNameManager.UpdateGenerateDNS(updateCtx, lookupTime, qname, &fqdn.DNSIPRecords{
+		IPs: responseIPs,
+		TTL: int(TTL),
+	})
+
+	if stat != nil {
+		stat.PolicyGenerationTime.End(true)
+		stat.DataplaneTime.Start()
+	}
+
+	if err := <-dpUpdates; err != nil {
+		log.Warning("Timed out waiting for datapath updates of FQDN IP information; returning response. Consider increasing --tofqdns-proxy-response-max-delay if this keeps happening.")
+		metrics.ProxyDatapathUpdateTimeout.Inc()
+	}
+
+	// Policy updates for this name have been pushed out; we can release the lock.
+	d.dnsNameManager.UnlockName(qname)
+
+	if logDebug {
+		log.WithFields(logrus.Fields{
+			logfields.Duration:   time.Since(updateStart),
+			logfields.EndpointID: ep.GetID(),
+			"qname":              qname,
+		}).Debug("Waited for endpoints to regenerate due to a DNS response")
+	}
+
+	return nil
+}
+
+// Standalone DNS Proxy grpc server is used to communicate with the Standalone DNS Proxy
+func (d *Daemon) bootstrapStandaloneDNSProxyServer() {
+	sdpServer := service.NewServer(d.endpointManager, d.updateOnDNSMsg, d.logger)
+	proxy.GlobalStandaloneDNSProxy = sdpServer
```

Comment:
nit: might be worth adding a comment here that this write to a global variable is not a data race because it is guaranteed to happen before endpoint regeneration, which calls `porxy.UpdateSDP` (which in turn reads `proxy.GlobalStandaloneDNSProxy`).

Reply:
This got removed in refactoring.